runtime-modules/project-token/src/lib.rs

```diff
@@ -0,0 +1,703 @@
+        // validate according to policy
+        outputs.iter().try_for_each(|(dst, _)| {
+            token_info.ensure_valid_location_for_policy::<T, T::AccountId, _>(dst)
+        })?;
+
+        // == MUTATION SAFE ==
+
+        Self::do_transfer(token_id, &src, outputs, decrease_operation);
+
+        let outputs_for_event = outputs
+            .iter()
+            .map(|(dst, amount)| (dst.location_account(), *amount));
+
+        Self::deposit_event(RawEvent::TokenAmountMultiTransferred(
+            token_id,
+            src,
+            outputs_for_event.collect(),
+        ));
+
+        Ok(())
+    }
+
+    /// Change to permissionless
+    /// Preconditions:
+    /// - Token `token_id` must exist
+    /// Postconditions
+    /// - transfer policy of `token_id` changed to permissionless
+    fn change_to_permissionless(token_id: T::TokenId) -> DispatchResult {
+        TokenInfoById::<T>::try_mutate(token_id, |token_info| {
+            token_info.transfer_policy = TransferPolicyOf::<T>::Permissionless;
+            Ok(())
+        })
+    }
+
+    /// Reduce patronage rate by amount
+    /// Preconditions:
+    /// - `token_id` must exists
+    /// - `decrement` must be less or equal than current patronage rate for `token_id`
+    ///
+    /// Postconditions:
+    /// - patronage rate for `token_id` reduced by `decrement`
+    fn reduce_patronage_rate_by(token_id: T::TokenId, decrement: Percent) -> DispatchResult {
+        let token_info = Self::ensure_token_exists(token_id)?;
+
+        // ensure new rate is >= 0
+        ensure!(
+            token_info.patronage_info.rate >= decrement,
+            Error::<T>::ReductionExceedingPatronageRate,
+        );
+
+        // == MUTATION SAFE ==
+
+        let new_rate = TokenInfoById::<T>::mutate(token_id, |token_info| {
+            let new_rate = token_info.patronage_info.rate.saturating_sub(decrement);
+            token_info.patronage_info.rate = new_rate;
+            new_rate
+        });
+
+        Self::deposit_event(RawEvent::PatronageRateDecreasedTo(token_id, new_rate));
+
+        Ok(())
+    }
+
+    /// Query for patronage credit for token
+    /// Preconditions
+    /// - `token_id` must exists
+    fn get_patronage_credit(token_id: T::TokenId) -> Result<T::Balance, DispatchError> {
+        Self::ensure_token_exists(token_id)
+            .map(|token_info| token_info.patronage_info.outstanding_credit)
+    }
+
+    /// Allow creator to receive credit into his accounts
+    /// Preconditions:
+    /// - `token_id` must exists
+    /// - `to_account` must be valid for `token_id`
+    ///
+    /// Postconditions:
+    /// - outstanding patronage credit for `token_id` transferred to `to_account`
+    /// - outstanding patronage credit subsequently set to 0
+    /// no-op if outstanding credit is zero
+    fn claim_patronage_credit(token_id: T::TokenId, to_account: T::AccountId) -> DispatchResult {
```

Comment:
I think this is a misunderstanding of how patronage is supposed to work.

Patronage is a rent as a function of _time_. This means, whenever time passes, `outstanding_credit` is effectively supposed to go up. From this, you can already see that we can exactly use the semantics that `outstanding_credit` represents exactly what this issuer is owner at any given time, because we would then have to recompute on every block. Instead, if you have it represent the cumulative unclaimed patronage amount as of the last block where either the patronage rate or the issuanc changed, then it works fine. You just need to store a block height of when this was, and then you can compute marginal patronage earned since that time at simpler per block rate.

You can reread the original spec on this

> The payout occurs through the owner initiating a cashout, flowing to their member account, which results in a suitable number of $CRT tokens being minted in this account. This calculation is straight forward, as it only needs to know when the last cashout, or change in issuance, occurs. Critically, if there is a change in issuance for any other reason as part of the system features (e.g. burning, or anything else), then there should should be a tally kept in the channel which reflects what would have been paid out to the owner if they had cashed out now, and then the timestamp for the cashout action is reset. 

or check out this issue which describes the same idea in the context of working group payouts being pull based

https://github.com/Joystream/joystream/issues/1305